core/common/src/main/java/alluxio/AbstractClient.java

```diff
@@ -168,8 +168,17 @@ public synchronized void connect() throws IOException {
 
     RetryPolicy retryPolicy =
         new ExponentialBackoffRetry(BASE_SLEEP_MS, MAX_SLEEP_MS, RPC_MAX_NUM_RETRY);
-    while (!mClosed) {
+    while (true) {
+      if (mClosed) {
+        throw new FailedPreconditionException("Failed to connect: client has been closed");
+      }
       mAddress = getAddress();
+      if (mAddress == null) {
+        if (retryPolicy.attemptRetry()) {
```

Comment:
done - I've updated the MasterInquireClient API to say that implementation should define their own retries. Now if the client returns null here, we fail without retrying any further.